apiserver/facades/agent/uniter/uniter.go

```diff
@@ -1448,54 +1450,78 @@ func (u *UniterAPI) ReadSettings(ctx context.Context, args params.RelationUnits)
-		}
+func (u UniterAPI) readOneUnitSettings(
+	ctx context.Context,
+	canAccessUnit common.AuthFunc,
+	arg params.RelationUnit,
+) (params.Settings, error) {
+	unitTag, err := names.ParseTag(arg.Unit)
+	if err != nil {
+		return nil, apiservererrors.ErrPerm
+	}
+	relTag, err := names.ParseRelationTag(arg.Relation)
+	if err != nil {
+		return nil, apiservererrors.ErrPerm
+	}
+	relUUID, err := u.relationService.GetRelationUUIDFromKey(ctx, corerelation.Key(relTag.Id()))
+	if errors.Is(err, errors.NotFound) {
+		return nil, apiservererrors.ErrPerm
+	} else if err != nil {
+		return nil, internalerrors.Capture(err)
+	}
 
-		var settings map[string]interface{}
+	var settings map[string]string
 
-		switch tag := tag.(type) {
-		case names.UnitTag:
-			var relUnit *state.RelationUnit
-			relUnit, err = u.getRelationUnit(canAccessUnit, arg.Relation, tag)
-			if err != nil {
-				return nil, errors.Trace(err)
-			}
-			var node *state.Settings
-			node, err = relUnit.Settings()
-			if err != nil {
-				return nil, errors.Trace(err)
-			}
-			settings = node.Map()
-		case names.ApplicationTag:
-			// Emulate a ReadLocalApplicationSettings call where
-			// the currently authenticated tag is implicitly
-			// assumed to be the requesting unit.
-			authTag := u.auth.GetAuthTag()
-			if authTag.Kind() != names.UnitTagKind {
-				// See LP1876097
-				return nil, apiservererrors.ErrPerm
-			}
-			settings, err = u.readLocalApplicationSettings(arg.Relation, tag, authTag.(names.UnitTag))
-		default:
+	switch tag := unitTag.(type) {
+	case names.UnitTag:
+		settings, err = u.readLocalUnitSettings(ctx, canAccessUnit, relUUID, tag)
+		if err != nil {
+			return nil, internalerrors.Capture(err)
+		}
+	case names.ApplicationTag:
+		// Emulate a ReadLocalApplicationSettings call where
+		// the currently authenticated wordpressUnitTag is implicitly
+		// assumed to be the requesting unit.
+		authTag := u.auth.GetAuthTag()
+		if authTag.Kind() != names.UnitTagKind {
+			// See LP1876097
 			return nil, apiservererrors.ErrPerm
 		}
+		settings, err = u.readLocalApplicationSettings(ctx, relUUID, tag, authTag.(names.UnitTag))
+	default:
+		return nil, apiservererrors.ErrPerm
+	}
 
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
-		return convertRelationSettings(settings)
+	if err != nil {
+		return nil, internalerrors.Capture(err)
 	}
+	return settings, nil
+}
 
-	for i, arg := range args.RelationUnits {
-		settings, err := readOneSettings(arg)
-		result.Results[i].Error = apiservererrors.ServerError(err)
-		result.Results[i].Settings = settings
+func (u *UniterAPI) readLocalUnitSettings(
+	ctx context.Context,
+	canAccess common.AuthFunc,
+	relUUID corerelation.UUID,
+	unitTag names.UnitTag,
+) (map[string]string, error) {
+	if !canAccess(unitTag) {
+		return nil, apiservererrors.ErrPerm
 	}
-	return result, nil
+
+	relUnitUUID, err := u.relationService.GetRelationUnit(ctx, relUUID, unitTag.Id())
+	if err != nil {
+		return nil, internalerrors.Capture(err)
+	}
```

Comment:
Do we need to handle the specific errors returned (or that will be returned in the future) by `GetRelationUnit()`?

Reply:
If we fail to get the relation unit uuid, there are no unique steps to be taken based on the error type. Just return the error.